### PR TITLE
Add `Sequence` protocol

### PR DESCRIPTION
**Pull request by MaxDesiatov**

### CODE REVIEW

File: Sources/Core/Sequence.swift [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Copyright © 2022 Max Desiatov <max@desiatov.com>.
+// All Rights Reserved.
+// SPDX-License-Identifier: BSD-3
+
+public protocol IteratorProtocol {
+  associatedtype Element
+
+  mutating func next() -> Element?
+}
```
Comment (compnerd) [inline Sources/Core/Sequence.swift:9]: Thoughts on splitting this up into two files?

Comment (MaxDesiatov) [reply Sources/Core/Sequence.swift:9]: Sure thing! I never found the stdlib tendency to group multiple things into same files very readable. OTOH could this lead to an issue where too many files exceed the CLI arguments limit with CMake? But we're too far away from that 😅

Comment (MaxDesiatov) [reply Sources/Core/Sequence.swift]: Done. LMK if you'd like this to be split into two separate PRs.

Comment (compnerd) [reply Sources/Core/Sequence.swift]: If we hit that limit, I'd actually prefer to fix CMake ;-).  Lets cross that bridge when we get to it.  Code is read far more than it is written, so I prefer that we optimize for search and reading.  As I said previously there is no desire for this to be a copy of the standard library.

**Review by compnerd — APPROVED** [review]

**Pull request merged**